### PR TITLE
openstack: support new authentication types

### PR DESCRIPTION
**Pull request by mmartinv**
Support the application credentials and token authentication types.

To choose one of the new authentication types a new field `authType`
is expected in the secret (defaults to `password` if not specified),
which can have one of the supported authentication types:

- password
- token
- v3password
- v3token
- v3applicationcredential

When choosing the application credentials authentication type
(`v3applicationcredentials`) three new secret fields are available:

- applicationCredentialID
- applicationCredentialName
- applicationCredentialSecret

One of `applicationCredentialID` or `applicationCredentialName`
is required alongside `applicationCredentialSecret`.

If `applicationCredentialName` is used it's also mandatory to
specify the `username` and `userDomainName` fields.

The application credentials must have a role assigned that permits to
read the required information (VMs, Flavors, Images, Volumes,
Snapshots, Networks...) and to stop and start VMs, (`reader` and
`member` are the typical ones). The authentication credentials
might be generated with the following command:

```
openstack application credential create \
          --role reader \
          --role member \
          --secret redhat forklift
```

When choosing the token authentication type (`token` or `v3token`)
the `projectID` or `projectName` and the `userID` or `username` are
required alongside the `token` field.

When using `projectName` it's also required to specify the `domainName`
field.

It's possible to generate the required token field with the following
command:

```
openstack token issue
```

Signed-off-by: Miguel Martín <mmartinv@redhat.com>

### CODE REVIEW

File: pkg/controller/provider/container/openstack/client.go [hunk 1]
```diff
@@ -30,6 +30,37 @@ import (
...
+	ProjectDomainName           = "projectDomainName"
+	ProjectDomainID             = "projectDomainID"
+	DomainName                  = "domainName"
+	DefaultDomain               = "defaultDomain"
+	InsecureSkipVerify          = "insecureSkipVerify"
+	CACert                      = "cacert"
+)
+
+var supportedAuthTypes = map[string]clientconfig.AuthType{
+	"password":                clientconfig.AuthPassword,
```
[the remark this answers is not in the document]
Comment (ahadas) [reply pkg/controller/provider/container/openstack/client.go:57]: yeah, and same for `v2token`

**Review by ahadas — COMMENTED** [review]
looks great, I also wonder about v2 of the password and token - they they intentionally not included? other than that, @eslutsky mentioned extending the CI to test the different authentication types and we also need to take into account the changes required in the UI - I'll add a ticket on this in Jira

**Comment by mmartinv** [comment]
@ahadas @bennyz Supporting identity V2 would require bigger changes in the code and I am not sure if it's still officially supported. Even Openstack 13 (Queens) comes with identity V3.
Right now we only support the latest major version for all the services: identity (v3), compute (v2), block storage (v3), image (v2) and network (v2). I am not sure if supporting identity V2 would require supporting other major versions of the rest of the services.

**Comment by ahadas** [comment]
@mmartinv if v3 is supported in Openstack 13 then I see no reason to support v2

**Review by ahadas — APPROVED** [review]

**Pull request merged**